galasa-extensions-parent/dev.galasa.auth.couchdb/src/main/java/dev/galasa/auth/couchdb/internal/CouchdbAuthStoreValidator.java

```diff
@@ -6,26 +6,176 @@
+
+    }
+    
+    public void checkTokensDesignDocument( CloseableHttpClient httpClient , URI couchdbUri , int attempts) throws CouchdbException {
+        HttpRequestFactory requestFactory = super.getRequestFactory();
+        HttpGet httpGet = requestFactory.getHttpGetRequest(couchdbUri + "/galasa_tokens/_design/docs");
+
+        String docJson = null;
+        try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
+
+            StatusLine statusLine = response.getStatusLine();
+
+            docJson = EntityUtils.toString(response.getEntity());
+            if (statusLine.getStatusCode() != HttpStatus.SC_OK
+                    && statusLine.getStatusCode() != HttpStatus.SC_NOT_FOUND) {
+                throw new CouchdbException(
+                        "Validation failed of database galasa_tokens design document - " + statusLine.toString());
+            }
+            if (statusLine.getStatusCode() == HttpStatus.SC_NOT_FOUND) {
+                docJson = "{}";
+            }
+        } catch (CouchdbException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new CouchdbException("Validation failed", e);
+        }
+
+        boolean updated = false;
+
+        JsonObject doc = gson.fromJson(docJson, JsonObject.class);
+        doc.remove("_id");
+        String rev = null;
+        if (doc.has("_rev")) {
+            rev = doc.get("_rev").getAsString();
+        }
+
+        JsonObject views = doc.getAsJsonObject("views");
+        if (views == null) {
+            updated = true;
+            views = new JsonObject();
+            doc.add("views", views);
+        }
+
+        JsonObject requestors = views.getAsJsonObject("loginId-view");
+        if (requestors == null) {
+            updated = true;
+            requestors = new JsonObject();
+            views.add("loginId-view", requestors);
+        }
+
+        if (checkView(requestors, "function (doc) { if (doc.owner && doc.owner.loginId) {emit(doc.owner.loginId, doc); } }", "javascript")) {
+            updated = true;
+        }
+
+        if (updated) {
+            logger.info("Updating the galasa_tokens design document");
+
+            HttpEntity entity = new StringEntity(gson.toJson(doc), ContentType.APPLICATION_JSON);
+
+            HttpPut httpPut = requestFactory.getHttpPutRequest(couchdbUri + "/galasa_tokens/_design/docs");
+            httpPut.setEntity(entity);
+
+            if (rev != null) {
+                httpPut.addHeader("ETaq", "\"" + rev + "\"");
+            }
+
+            try (CloseableHttpResponse response = httpClient.execute(httpPut)) {
+                StatusLine statusLine = response.getStatusLine();
+                int statusCode = statusLine.getStatusCode();
+                if (statusCode == HttpStatus.SC_CONFLICT) {
+                    // Someone possibly updated
+                    attempts++;
+                    if (attempts > 10) {
```

Comment:
10 should be a symbolic like MAX_COUCHDB_CONFLICT_RETRIES_BEFORE_GIVE_UP ?

Reply:
Also surprised that this backoff-time isn't all over the couchdb code, and collected into a single place.